refactor(gossipsub): send more specific messages to `ConnectionHandler`

jxs opened the pull request:
## Description

Previously, the `NetworkBehaviour` constructed a `proto::RPC` type and sent that into an unbounded queue to the `ConnectionHandler`. With such a broad type, the `ConnectionHandler` cannot do any prioritization on which messages to drop if a connection slows down.

To enable a more fine-granular handling of messages, we make the interface between `NetworkBehaviour` and `ConnectionHandler` more specific by introducing an `RpcOut` type that differentiates between `Publish`, `Forward`, `Subscribe`, etc.

Related: #4667.

## Notes & open questions
Max suggested on https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1788658203 changing `HandlerIn` but that would imply more changes into the `fragment_message` function and back and forth conversion between `proto::RPC` and its subtypes (using a subtype via `HandlerIn` to then convert to the main `proto::RPC` type when sending it from the `ConnectionHandler`). Adding a `RpcType` to distinguish seemed simpler.

We were also using the `Rpc` type for conversions between `Messages`, `ControlActions`, and `Subscriptions` which can be done from these types directly.

ptal @mxinden @thomaseizinger @divagant-martian @AgeManning 

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

jxs commented on protocols/gossipsub/src/lib.rs:
```diff
@@ -125,7 +125,7 @@ pub use self::subscription_filter::{
 };
 pub use self::topic::{Hasher, Topic, TopicHash};
 pub use self::transform::{DataTransform, IdentityTransform};
-pub use self::types::{Message, MessageAcceptance, MessageId, RawMessage, Rpc};
```
`Rpc` doesn't need to be public right?

AgeManning replied:
I dont think so

thomaseizinger replied:
I thought it is mentioned in some transform API or something? You should be able to test by making it `pub(crate)`.

Note that that is a breaking change so we can't do it right away.

AgeManning replied:
oh yeah good point. The compression algo prob needs it.

thomaseizinger replied:
I just tried and it can be made private! :tada:

thomaseizinger replied:
> The compression algo prob needs it.

`RawMessage` could be private too if we'd deprecate `DataTransform` in favor of an `fn(Message) -> Message` right? Would that be an option or do you necessarily need the data that is in `RawMessage` but not `Message`?

jxs commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -3702,20 +3640,32 @@ mod local_test {
-            }
-            for _ in 0..g.gen_range(0..10u8) {
-                rpc.control_msgs.push(test_control());
+            match u8::arbitrary(g) % 3 {
```
we only send one type of gossipsub message so I'd say this reflects better on our usage

jxs commented on protocols/gossipsub/src/behaviour/tests.rs:
```diff
@@ -660,7 +669,11 @@ fn test_publish_without_flood_publishing() {
-                event: HandlerIn::Message(ref message),
+                event:
+                    HandlerIn::Message {
+                        rpc_type: RpcType::Forward,
```
this particular one (`test_publish_without_flood_publishing`) should instead be `RpcType::Publish` right? asking cause the method called by the `NetworkBehaviour` is `forward_msg`

jxs committed:
separate RpcOut from Rpc,

rename Rpc to RpcIn and create RpcOut to allow sending different types of messages via such enum.
this will alow us separing into different priorities the types of messages.

jxs committed:
remove fragment_message function,

it doesn't fragment Publish/Forward messages as it's not on the spec data fragmentation, and separating
ControlAction and Subscription message optimization should only be added with proofs
(calculation/benchmarks) it improves performance. Removing it will allow us to simplify the code by
only converting to proto::RPC on the connection handler.

jxs committed:
send RpcOut to connection handler

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -576,15 +571,10 @@ where
+            let event = RpcOut::Subscriptions(vec![Subscription {
+                topic_hash: topic_hash.clone(),
+                action: SubscriptionAction::Unsubscribe,
+            }]);
```
I would like to remove as many of the nested lists as possible. If we ever need to send multiple subscriptions at once somewhere (here we don't seem to be?), then we can just send multiple events instead.

thomaseizinger replied:
We could even simplify this further by making `RpcOut` have `Subscribe` and `Unsubscribe` variants, then we don't need to pass the `SubscriptionAction` down.

In general, I think it will lead to much easier to understand code if we move towards flat hierarchies here that avoid nested matching and loops.

jxs replied:
yup, I like that also, thanks Thomas! Updated

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -638,10 +636,6 @@ where
 
-        // If we are not flood publishing forward the message to mesh peers.
-        let mesh_peers_sent = !self.config.flood_publish()
-            && self.forward_msg(&msg_id, raw_message.clone(), None, HashSet::new())?;
```
Why is this safe to remove? Does `forward_msg` not have a side-effect that we are now no longer performing?

jxs replied:
thanks for raising this Thomas, I was going to annotate it but you went ahead first :grin: 
basically `forward_msg` repeats all the peer counting that is done below besides the mesh peers, so I took that part of code and put it into the logic bellow when flood subbing is not enabled, this way we save repeated messages and are able to use `forward_msg` for `RpcOut::Forward`

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -576,15 +571,10 @@ where
         // announce to all peers
         let peer_list = self.peer_topics.keys().cloned().collect::<Vec<_>>();
         if !peer_list.is_empty() {
```
It is not related to this PR but what is the point of this check? A loop with no entries is a no-op so we don't need to check this, right?

If you don't mind a bit of clean-up while we are at it, you could send a separate PR for tiny refactorings like this.

Note that as is, the code does have a slight performance optimisation in that we don't allocate if we the list is empty. But we can alleviate that be inlining the construction of the event into the loop and thus never perform it when the list is empty.

jxs replied:
yup makes sense Thomas, did it in f2c5edd766f5ad0153ef292b7e150013ddfb970b

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -1360,14 +1348,11 @@ where
-                control_msgs: Vec::new(),
-            }
-            .into_protobuf();
+            let msg_bytes = message_list
```
Similarly as above, I think we should directly queue a single event `Forward` with a single message in the loop instead of collecting in `cached_messages` first and then again (kind of unnecessarily) check whether it is empty and build a single RPC out of it.

Now looking at this code, I don't believe the metrics below are correct. We seem to be incrementing the bytes of _all_ messages for _each_ topic which in total will be a multiple of what we actually transferred.

jxs replied:
yeah makes sense, updated.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2867,146 +2809,21 @@ where
     }
 
-    /// Send a [`Rpc`] message to a peer. This will wrap the message in an arc if it
+    /// Send a [`RpcOut`] message to a peer. This will wrap the message in an arc if it
```
Stale docs here.

jxs replied:
updated!

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2867,146 +2809,21 @@ where
+    /// Send a [`RpcOut`] message to a peer. This will wrap the message in an arc if it
     /// is not already an arc.
-    fn send_message(&mut self, peer_id: PeerId, message: proto::RPC) -> Result<(), PublishError> {
+    fn send_message(&mut self, peer_id: PeerId, message: RpcOut) -> Result<(), PublishError> {
```
Perhaps not in this PR but the metrics and logging of what we are doing should be moved to this function in order to de-duplicate it from all the call-sites.

jxs replied:
did it on an isolated commit: fdee626aa873fea736b2e6dde05c22ebc4e14147

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2867,146 +2809,21 @@ where
     /// is not already an arc.
-    fn send_message(&mut self, peer_id: PeerId, message: proto::RPC) -> Result<(), PublishError> {
+    fn send_message(&mut self, peer_id: PeerId, message: RpcOut) -> Result<(), PublishError> {
         // If the message is oversized, try and fragment it. If it cannot be fragmented, log an
```
Stale comment.

jxs replied:
thanks, updated

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -3071,15 +2888,7 @@ where
-                            }
-                            .into_protobuf(),
-                        )
+                        .send_message(peer_id, RpcOut::Subscriptions(subscriptions))
```
Same as above, instead of collecting in a temporary `Vec`, lets call `self.send_message` for each subscription that is pending.

thomaseizinger replied:
Another refactoring we can do here is to be more defensive with our `if`s:

```rust
if other_established > 0 {
	return; // Not our first connection to this peer, hence nothing to do.
}

if self.blacklisted_peers.contains(&peer_id) {
	tracing::debug!(peer=%peer_id, "Ignoring connection from blacklisted peer");
	return;
}
```

YMMV but I much prefer this style because it reduces nesting and allows my brain to think through these conditions individually instead of all at once.

thomaseizinger replied:
Ideally, we would move the sending of pending subscriptions to the `handle_established_{in,out}bound_connection` callback and directly give them to the handler. That way, it can immediately start processing these messages instead of injecting them here in the callback.

This wasn't as easily possible when this code was written because the previous `new_handler` function had an indirection until it created the `ConnectionHandler`.

This is definitely a refactoring that we can do at a later point though.

jxs replied:
yeah, agree and updated, but the second case you mention I also noticed and I agree that we should return earlier but that is different behaviour than what we currently have ptal Thomas, did it on an isolated commit:  f2c5edd766f5ad0153ef292b7e150013ddfb970b

thomaseizinger replied:
> but that is different behaviour than what we currently have

You can restore the old behaviour but just moving everything that needs to always happen before the early return :)
That is nice thing about this style, it makes it explicit that something happens independently of a certain condition instead of having to parse in your head, where certain code branches add and which other conditions are essentially implicit.

thomaseizinger reviewed (commented):
Thanks for pushing this forward @jxs!

I've left a lot of ideas in what we can do here to make things simpler. We don't need to tackle all of them but it would be nice if we could tidy things up a bit when we are already touching the gossipsub code :)

jxs committed:
send_message doesn't need to return Result anymore

jxs committed:
split RpcOut::Subscriptions

into RpcOut::Subscribe and RpcOut::Unsubscribe. Send multiple Subscribe messages instead of a single
one with a subscription list.

jxs committed:
send multiple Forward message RpcOut's,

instead of a single Forward RpcOut with multiple messages

jxs committed:
deduplicate metrics accounting into send_message

jxs committed:
small refactorings to make the code more readable

jxs reviewed (commented):
Thanks for the review Thomas! Addressed the changes in separated commits, it's easier if you follow them by order.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2817,24 +2813,27 @@ where
-            }
+        tracing::debug!(peer=%peer_id, "New peer connected");
+        // We need to send our subscriptions to the newly-connected node.
+        for topic_hash in self.mesh.keys().cloned().collect::<Vec<_>>() {
```
I don't think the `collect` here is necessary?

jxs replied:
yeah it is because of the split borrows again

thomaseizinger commented on protocols/gossipsub/src/behaviour/tests.rs:
```diff
@@ -402,26 +402,16 @@ fn test_subscribe() {
         .events
         .iter()
-        .fold(vec![], |mut collected_subscriptions, e| match e {
+        .fold(0, |collected_subscriptions, e| match e {
```
This could now be a `.filter` with `matches!` and a `.count` probably.

jxs replied:
indeed, thanks! Updated.

thomaseizinger commented on protocols/gossipsub/src/protocol.rs:
```diff
@@ -607,7 +602,7 @@ mod tests {
                 HandlerEvent::Message { mut rpc, .. } => {
                     rpc.messages[0].validated = true;
 
-                    assert_eq!(rpc, rpc);
```
Lol, I don't think that assertion previously did anything because the `rpc` from the `match` arm shadows the one created on line 595 :upside_down_face:

jxs replied:
yeah exactly ehe

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -536,20 +536,11 @@ where
         // send subscription request to all peers
         let peer_list = self.peer_topics.keys().cloned().collect::<Vec<_>>();
         if !peer_list.is_empty() {
```
I think we can get rid of the `.collect` here to and just iterate over `self.peer_topics`.

jxs replied:
yeah it should but it doesn't, we can't because of the split borrows rule.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -574,22 +565,10 @@ where
-                tracing::debug!(%peer, "Sending UNSUBSCRIBE to peer");
-                self.send_message(peer, event.clone())?;
-            }
+        for peer in self.peer_topics.keys().cloned().collect::<Vec<_>>() {
```
```suggestion
        for (peer, _) in self.peer_topics {
```

Does this work?

jxs replied:
no for the same reason ehe, split borrows we borrow `self` as immutable with the iteration which impedes us from borrowing as mutable on `send_message`

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -574,22 +565,10 @@ where
-                self.send_message(peer, event.clone())?;
-            }
+        for peer in self.peer_topics.keys().cloned().collect::<Vec<_>>() {
+            tracing::debug!(%peer, "Sending UNSUBSCRIBE to peer");
```
Could move the logging into `send_message` too :)

jxs replied:
yeah I thought about that, but we're only calling it once here, and in `send_message` we'd need to match against the Rpc variants

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -1350,33 +1322,10 @@ where
-                    m.msg_sent(topic, msg_bytes);
-                }
-            }
+            self.send_message(*peer_id, RpcOut::Forward(message));
```
Can we move this call to line 1320 and thus eliminate the need for `cached_messages`?

jxs replied:
yeah, done! Thanks for this Thomas, I was going to do it on a subsequent PR I am thankful you went ahead and took a look :bow:

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -1536,18 +1485,7 @@ where
-            ) {
-                tracing::error!("Failed to send PRUNE: {:?}", e);
-            }
+            self.send_message(*peer_id, RpcOut::Control(prune_messages));
```
Same as with the other ones, I think it would be easier to understand if we would queue each individual `prune` message as a separate `RpcOut`.

jxs replied:
yup, addressed!

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2721,20 +2626,11 @@ where
 
         // forward the message to peers
         if !recipient_peers.is_empty() {
```
Instead of building up `recipient_peers`, can we just call `send_message` every time we are currently inserting a peer into `recipient_peers`?

jxs replied:
I'd say that gets more complex as we have multiple conditions which would then become different loops

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -668,6 +638,13 @@ where
                 );
             } else {
+                // Mesh peers
+                if let Some(mesh_peers) = self.mesh.get(&raw_message.topic) {
```
Could be an `unwrap_or_default` instead of `if let`, paired with an `.extend`.

thomaseizinger replied:
Line 667 seems to be the opposite of this here, should be perhaps combine this into a `match` instead?

jxs replied:
yup makes sense, Thanks Thomas

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -624,18 +603,13 @@ where
         }
 
+        let event = RpcOut::Publish(raw_message.clone());
+
```
I think I'd prefer constructing `RpcOut` as late as possible, perhaps even inline with `send_message`?

jxs replied:
yeah agreed, updated Thomas.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2721,20 +2626,11 @@ where
-                if let Some(m) = self.metrics.as_mut() {
-                    m.msg_sent(&message.topic, msg_bytes);
-                }
+                self.send_message(*peer, event.clone());
```
It might almost be worth it to change `send_message` to take an `impl IntoIterator<Item = PeerId>` and do the `for` internally?

jxs replied:
yeah I tried that, but we still have the split borrows problem

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -2847,164 +2743,28 @@ where
     /// Takes each control action mapping and turns it into a message
     fn flush_control_pool(&mut self) {
```
This probably exceeds the scope of this refactoring but I am curious to understand if the holding back of these messages is that important?

If we are now adding prioritization to the sending of messages, would it make sense to remove this pool and instead send them with a lower priority?

jxs replied:
yeah makes sense, can we leave this to the prioritization PR?

thomaseizinger reviewed (commented):
Couple more suggestions, most of them pretty minor. Feel free to not address immediately / at all :)

Overall, I think this is already looking great and much easier to understand! I wouldn't be opposed to merge this as an initial refactoring and start another PR that explores how we can add a (prioritised) channel to the handler.

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -231,125 +244,151 @@ pub struct Rpc {
+        }
+        b.finish()
+    }
+}
```
Can we move this to the bottom of the file again to avoid the diff?

jxs replied:
yeah, right I noticed that and thought had addressed, updated thanks!

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -231,125 +244,151 @@ pub struct Rpc {
+impl RpcOut {
     /// Converts the GossipsubRPC into its protobuf format.
     // A convenience function to avoid explicitly specifying types.
     pub fn into_protobuf(self) -> proto::RPC {
```
semver-checks complains that this function is now removed from the public `Rpc` type. You'll have to add it back to make this backwards compatible :)

jxs replied:
ehe right, makes sense. Updated, want to add a deprecation note?

thomaseizinger replied:
I don't think we need to with https://github.com/libp2p/rust-libp2p/pull/4833 on the way :)

jxs replied:
ah nice! Thanks Thomas

jxs committed:
address review

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -220,6 +233,122 @@ pub enum ControlAction {
+                            topic_id: Some(topic_hash.into_string()),
+                            message_ids: message_ids.into_iter().map(|msg_id| msg_id.0).collect(),
+                        };
+                        control.ihave.push(rpc_ihave);
```
You can directly `return` here now with a single `proto::RPC` message.

jxs replied:
thanks!

thomaseizinger commented on protocols/gossipsub/src/types.rs:
```diff
@@ -220,6 +233,122 @@ pub enum ControlAction {
+                    prune: Vec::new(),
+                };
+
+                match action {
```
I'd suggest to merge this `match` in the outer one and have `RpcOut::Control(ControlAction::IHave { .. })` etc.

jxs replied:
yup becomes nittier, thanks, updated!

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -668,6 +632,51 @@ where
+                    None => {
+                        tracing::debug!(topic=%topic_hash, "Topic not in the mesh");
+                        // If we have fanout peers add them to the map.
+                        if self.fanout.contains_key(&topic_hash) {
```
This could be a `match`-arm guard of `None` which would reduce indentation. Not sure if that is necessarily better than an `if else`.

jxs replied:
there's still code after the if which would then need to be repeated for each if else

thomaseizinger replied:
Yeah fair. It is only a single line so I wouldn't mind too much but happy to leave it as is.

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -1331,50 +1298,24 @@ where
                         peer=%peer_id,
                         message=%id,
                         "IWANT: Peer has asked for message too many times; ignoring request"
                     );
                 } else {
```
I'd prefer `continue` here over an `else`.

jxs replied:
as the above this also has another if else after this one that would need to be repeated here

thomaseizinger reviewed (commented):
Great stuff! Thanks for following up :)

Mind writing up what we discussed regarding a priority channel in the issue next?

jxs reviewed (commented):
Thanks Thomas! Wrote https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1808578183 as a follow up

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into gossipsub-backpressure-1

jxs committed:
cargo clippy

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into gossipsub-backpressure-1

jxs committed:
remove unrequired clones

and rename cloned to copied for clarity

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into gossipsub-backpressure-1

thomaseizinger reviewed (approved):
Great work! Happy to merge that as is :)
